### ISSUE TITLE
⚙Tech Debt: Consider introducing more explicit types

### ISSUE THREAD

**Issue by voxpelli** [melt-ui/melt-ui#978]
Currently a lot of the types in `melt-ui` seems to be inferred from the implementation rather than being explicitly typed. This has a tendency to create much more complex types than what an explicitly defined type would do + it makes it more difficult to refer to the types from elsewhere without oneself relying on something like `ReturnType<typeof createTabs>["elements"]["trigger"]`

Here's an example of [the generated type of `createTabs`](https://unpkg.com/@melt-ui/svelte@0.73.0/dist/builders/tabs/create.d.ts) in the latest version, `0.73.0`:

<details>
<summary>58 line long return type definition</summary>

```ts
/// <reference types="svelte" />
import type { MeltActionReturn } from '../../internal/types.js';
import type { TabsEvents } from './events.js';
import type { CreateTabsProps, TabsTriggerProps } from './types.js';
export declare function createTabs(props?: CreateTabsProps): {
    elements: {
        root: import("../../internal/helpers/index.js").MeltElement<import("../../internal/helpers/index.js").WithGet<import("svelte/store").Writable<import("../../internal/types.js").Orientation>>, import("svelte/action").Action<any, any, Record<never, any>>, ($orientation: import("../../internal/types.js").Orientation) => {
            'data-orientation': import("../../internal/types.js").Orientation;
        }, string>;
        list: import("../../internal/helpers/index.js").MeltElement<import("../../internal/helpers/index.js").WithGet<import("svelte/store").Writable<import("../../internal/types.js").Orientation>>, import("svelte/action").Action<any, any, Record<never, any>>, ($orientation: import("../../internal/types.js").Orientation) => {
            role: string;
            'aria-orientation': import("../../internal/types.js").Orientation;
            'data-orientation': import("../../internal/types.js").Orientation;
        }, string>;
        trigger: import("../../internal/helpers/index.js").MeltElement<[{
            update: (updater: import("svelte/store").Updater<string>, sideEffect?: ((newValue: string) => void) | undefined) => void;
            set: (this: void, value: string) => void;
            subscribe(this: void, run: import("svelte/store").Subscriber<string>, invalidate?: import("svelte/store").Invalidator<string> | undefined): import("svelte/store").Unsubscriber;
            get: () => string;
            destroy?: (() => void) | undefined;
        }, import("../../internal/helpers/index.js").WithGet<import("svelte/store").Writable<import("../../internal/types.js").Orientation>>], (node: HTMLElement) => MeltActionReturn<TabsEvents['trigger']>, ([$value, $orientation]: [string, import("../../internal/types.js").Orientation]) => (props: TabsTriggerProps) => {
            type: "button";
            role: string;
            'data-state': string;
            tabindex: number;
            'data-value': string;
            'data-orientation': import("../../internal/types.js").Orientation;
            'data-disabled': boolean | undefined;
            disabled: boolean | undefined;
        }, string>;
        content: import("../../internal/helpers/index.js").MeltElement<{
            update: (updater: import("svelte/store").Updater<string>, sideEffect?: ((newValue: string) => void) | undefined) => void;
            set: (this: void, value: string) => void;
            subscribe(this: void, run: import("svelte/store").Subscriber<string>, invalidate?: import("svelte/store").Invalidator<string> | undefined): import("svelte/store").Unsubscriber;
            get: () => string;
            destroy?: (() => void) | undefined;
        }, import("svelte/action").Action<any, any, Record<never, any>>, ($value: string) => (tabValue: string) => {
            role: string;
            'aria-labelledby': string;
            hidden: boolean | undefined;
            tabindex: number;
        }, string>;
    };
    states: {
        value: {
            update: (updater: import("svelte/store").Updater<string>, sideEffect?: ((newValue: string) => void) | undefined) => void;
            set: (this: void, value: string) => void;
            subscribe(this: void, run: import("svelte/store").Subscriber<string>, invalidate?: import("svelte/store").Invalidator<string> | undefined): import("svelte/store").Unsubscriber;
            get: () => string;
            destroy?: (() => void) | undefined;
        };
    };
    options: import("../../internal/helpers/index.js").ToWritableStores<Omit<{
        defaultValue?: string | undefined;
        value?: import("svelte/store").Writable<string> | undefined;
        onValueChange?: import("../../internal/helpers/index.js").ChangeFn<string> | undefined;
        orientation: import("../../internal/types.js").Orientation;
        activateOnFocus: boolean;
        loop: boolean;
        autoSet: boolean;
    }, "value" | "onValueChange" | "defaultValue" | "autoSet">>;
};
```

</details>

This is in turn referenced when defining these types:

```ts
export type Tabs = BuilderReturn<typeof createTabs>;
export type TabsElements = Tabs['elements'];
export type TabsOptions = Tabs['options'];
export type TabsStates = Tabs['states'];
```

If I want to refer to the type of `TabsElements['triggers']` or make a generic helper for such a trigger, then my only choice is to use the exact `TabsElements['triggers']` type.

Preferably the return type of `createTabs` should be the `Tabs` type from `types.ts` instead of the other way around:

```ts
type Tabs = {
	elements: TabsElements,
	states: TabsStates,
	options: TabsOptions,
};
```

And eg. `TabsElement` should be something like:

```ts
interface TabsElement {
	root: TabsRootElement,
	list: TabsListElement,
	trigger: TabsTriggerElement,
	content: TabsContentElement,
} 
```

Then I can simply do something like `import type { TabsTriggerElement } from '@melt-ui/svelte'; ` when I want to / need to refer to the `TabsTriggerElement` and not have it be dissected through the `typeof` of the actual creation function (this is my use case but something that the intellisense of eg. VSCode gives me no help in achieving as it has no knowledge that whatever `createTabs` returns can be referenced through eg. `TabsElement['trigger']` already).

It may even be preferable to double down on this and have the interfaces extend an interface it shares with other trigger-able  elements:

```ts
interface MeltElementWithTrigger {
	root: MeltElementRoot,
	trigger: MeltElementTrigger,
	content: MeltElementContent,
} 
interface TabsElement extends MeltElementWithTrigger {
	list: TabsListElement,
} 
```

That way one could easily refer to a `MeltElementTrigger` and know they all refer to the same kind of implementation (if that's the case).

Some references to back up these thoughts:

The [Writing Easy-to-Compile Code](https://github.com/microsoft/TypeScript/wiki/Performance#writing-easy-to-compile-code) section of the [TypeScript performance](https://github.com/microsoft/TypeScript/wiki/Performance#writing-easy-to-compile-code) wiki page mentions:

* Using Type Annotations:
  
  > Adding type annotations, especially return types, can save the compiler a lot of work. In part, this is because named types tend to be more compact than anonymous types [...] Type inference is very convenient, so there's no need to do this universally - however

* Preferring Interfaces Over Intersections:
  
  > Interfaces create a single flat object type that detects property conflicts, which are usually important to resolve! Intersections on the other hand just recursively merge properties, and in some cases produce never. Interfaces also display consistently better, whereas type aliases to intersections can't be displayed in part of other intersections. Type relationships between interfaces are also cached, as opposed to intersection types as a whole. A final noteworthy difference is that when checking against a target intersection type, every constituent is checked before checking against the "effective"/"flattened" type.
  > 
  > For this reason, extending types with interfaces/extends is suggested over creating intersection types.

I think its also a matter of typing philosophy – should types merely be a mirroring of the implementation or should types be an explicit contract that concisely defines what API one is offered.

I think explicit contracts that concisely defines what API:s one are offered are preferable and more useful.

And to refer to your contribution guidelines on Heuristics:

> * Complexity should be introduced when it’s inevitable
> * Code should be easy to reason about
> * Code should be easy to delete
> * Avoid abstracting too early

Depending on how one define `Complexity` it may or may not be considered more complex to have explicit rather than implicit types. When it comes to `reason` about the code though, explicit concise types are much easier to reason about whereas inferred ones can be very hard to reason about.

Is it easier to `delete`? I would say it's both yes and no – its easier to know what types are explicitly used by every function and thus easier to delete, but it is more lines of code so one could argue that its harder. I think its a negligible difference.

`Avoid abstracting too early` – being explicit about types can be considered an abstraction and can be deemed as something that is unnecessary / too early to do but I think that the explicit vs implicit dimension of code ultimately is a different dimension to the abstraction one – one can have abstracted code that's still highly implicit about its conditions (in a way the code that's in melt-ui for creating these elements right now) and one can have very non-abstracted code that's highly explicit about its conditions.

I would say that being explicit helps in understanding the nature of an abstraction and as such could possibly be considered a prerequisite for abstractions, but it's ultimately a different aspect of the code to the act of abstraction itself.

**Comment by TGlide** [comment]
Duplicate #510

**Comment by TGlide** [comment]
I want to give a clearer response though than merely say duplicate! Give me a moment,

**Comment by voxpelli** [comment]
👍 If anything I think that issue should be a duplicate of this one as `clearer types` is a somewhat ambiguous request 😜

**Comment by TGlide** [comment]
First of all, thank you for filling out this detailed issue. I'll reopen this one and close #510 as it's much better detailed, I agree with you.

This is something I do want to explore. I want to figure out a way to make the authoring experience still manageable when declaring these explicit types, but in the end, I concur that what matters most is the simplicity and discoverability of these exported types.

> I think its also a matter of typing philosophy – should types merely be a mirroring of the implementation or should types be an explicit contract that concisely defines what API one is offered.
>
>I think explicit contracts that concisely defines what API:s one are offered are preferable and more useful.

Maybe so. But then questions arise, should I be super detailed over what these types entail? With inferring, you get useful information that can be passed, for example, to Svelte's language server. E.g. if one of Melt's stores passes in a `type="button"` to a non-button element, it (should) complain. Typing these in manually is unrealistic, I'd say.

But it's debatable whether this information is *that* useful. Need to explore how I'd work around these.

One API I flirt with, is making the `create` functions use a generic creator function. Something like `makeBuilder`, which enforces some strict return types. It could be useful here, by *requiring* us to pass in other types as generics.

**Comment by TGlide** [comment]
P.S. Sorry for prematurely closing without much of a response! It's my anxiety screaming at me when I see the number of issues growing and me wanting to keep them down as much as possible 😄

**Comment by voxpelli** [comment]
Yeah, a generic creator function could potentially go far and I would also explore the composability of types and whether it would be feasible to set up and maintain something like:

```ts
interface MeltElementWithTrigger {
	root: MeltElementRoot,
	trigger: MeltElementTrigger,
	content: MeltElementContent,
}

interface TabsElement extends MeltElementWithTrigger {
	list: TabsListElement,
} 
```

**Comment by TGlide** [comment]
> One potentially low hanging fruit by the way would be to modify eg `MeltElement` and give it default types:
> 
> ```ts
> export type MeltElement<
> 	S extends Stores | undefined = undefined,
> 	// eslint-disable-next-line @typescript-eslint/no-explicit-any
> 	A extends Action<any, any> = Action<any, any>,
> 	R extends ElementCallback<S> = ElementCallback<S>,
> 	Name extends string = string
> > = MeltElementStore<S, A, R, Name> & A;
> ```
> 
> That way one can reference a `MeltElement` wherever one expects a `MeltElement` and ideally it should accept all kinds of such elements.
> 
> It would also enable eg. `satisfies MeltElement`

We do export `AnyMeltElement`, but that could be an improvement 🤔

**Comment by voxpelli** [comment]
Right, I see now, yeah I think providing defaults is pretty common and makes it more discoverable as then one can find that type simply by drilling deeper into the type, one never has to find a sibling or child type to use instead

**Comment by TGlide** [comment]
I'm playing around with auto-generating API reference from JSDoc, and I feel tackling this would be of huge help. Maybe I'll work on them simultaneously

**Comment by anxpara** [comment]
big +1 to this issue

right now I had to write

```ts
const sharedOptionsBoxesByName = new Map<string, any>(
    sharedOptionNames.map((name) => [
      name,
      createCheckbox({
        defaultChecked: false,
      }),
    ]),
  );
```

i'll try ```ReturnType<typeof createCheckbox>```, but still not ideal

**Comment by anxpara** [comment]
Now I'm trying to use the mapped Melt Checkboxes in an {#each}, but i'm hitting a few errors depending what i try

```
{#each optionNames as name}
  <button
      use:melt={???????}
      class="check-button"
      class:checked={???????}
      id="checkbox"
  >
    <input use:melt={$input} />
  </button>
{/each}
```

focusing on use:melt...

svelte doesn't let you use the ! operator inlined in dom, so i have to create at least one function to assert the option name exists in the map

```ts
type MeltCheckbox = ReturnType<typeof createCheckbox>;

function getOptionBox(name: OptionName): MeltCheckbox {
  return optionBoxesByName.get(name)!;
}

// trying to create a function to extract just the root store was a typing mess,
// and svelte also complained about only creating stores at the top level
```

svelte wouldn't let me do the following, cause the dollar tries to apply to the function

```
use:melt={$getOptionBox(name).elements.root}
```

this compiles, but is not reactive / doesn't set on click

```
use:melt={get(getOptionBox(name).elements.root)}
class:checked={get(getOptionBox(name).states.checked)}
```

i'll follow up in the discord support channel
